File: greedy_algorithm/jump_game_II.py
def jump(nums) -> int:
    i,curr,length,next = 0,-1, len(nums) - 1, -1
    ans = 0
    while next < length:
        if i > curr:
            ans += 1
            curr = next
        next = max(next, i + nums[i])
        i+=1
    return ans

# nums = [2,3,1,1,4] #ok
# nums = [2,3,0,1,4] #ok
nums = [0]
nums = [3,2,1]
print(jump(nums))

File: greedy_algorithm/_go/jump_game_II.go
package main

import "math"

func jump(nums []int) int {
	i, curr, next, length, ans := 0, -1, 0, len(nums)-1, 0
	for next < length {
		if i > curr {
			ans++
			curr = next
		}
		next = int(math.Max(float64(next), float64(i+nums[i])))

	}
	return ans
}
func main() {

}
